test/integration/tcp_tunneling_integration_test.cc

```diff
@@ -105,6 +107,50 @@ TEST_P(ConnectTerminationIntegrationTest, OriginalStyle) {
   sendBidirectionalDataAndCleanShutdown();
 }
 
+// Test Envoy CONNECT termination feature works with dynamic forward proxy configuration.
+TEST_P(ConnectTerminationIntegrationTest, OriginalStyleWithDynamicForwardProxy) {
+  // Add dynamic forward proxy filter into HTTP connection manager.
+  const std::string filter = fmt::format(R"EOF(
+name: dynamic_forward_proxy
+typed_config:
+  "@type": type.googleapis.com/envoy.extensions.filters.http.dynamic_forward_proxy.v3.FilterConfig
+  dns_cache_config:
+    name: foo
+    dns_lookup_family: {}
+)EOF",
+                                         Network::Test::ipVersionToDnsFamily(GetParam().version));
+  config_helper_.prependFilter(filter);
+
+  // Add dynamic forward proxy cluster config into cluster.
+  config_helper_.addConfigModifier([&](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    auto* cluster = bootstrap.mutable_static_resources()->mutable_clusters(0);
+    cluster->set_lb_policy(envoy::config::cluster::v3::Cluster::CLUSTER_PROVIDED);
+    cluster->mutable_cluster_type()->set_name("envoy.clusters.dynamic_forward_proxy");
+    envoy::extensions::clusters::dynamic_forward_proxy::v3::ClusterConfig cluster_config;
+    cluster_config.set_allow_insecure_cluster_options(true);
+    cluster_config.mutable_dns_cache_config()->set_name("foo");
+    // Set the dynamic forward proxy cluster dns_lookup_family based on input IpVersion.
+    envoy::config::cluster::v3::Cluster::DnsLookupFamily dns_lookup_family;
+    if (GetParam().version == Network::Address::IpVersion::v4) {
+      dns_lookup_family = envoy::config::cluster::v3::Cluster::V4_ONLY;
+    } else {
+      dns__lookup_family = envoy::config::cluster::v3::Cluster::V6_ONLY;
+    }
+    cluster_config.mutable_dns_cache_config()->set_dns_lookup_family(dns_lookup_family);
+    cluster->mutable_cluster_type()->mutable_typed_config()->PackFrom(cluster_config);
+  });
+
+  initialize();
+  clearExtendedConnectHeaders();
+  // Set the :authority filed in a request to be 'localhost' so it can be resolved by
+  // dynamic forward proxy DNS functionality.
+  connect_headers_.removeHost();
+  connect_headers_.setHost(
+      fmt::format("localhost:{}", fake_upstreams_[0]->localAddress()->ip()->port()));
```

Comment:
Don't you want to set the CONNECT URI rather than special host?

Reply:
With the DFP configured in HCM and cluster, Envoy will perform DNS resolution for the CONNECT target, which is the :authority field in the connect_headers_ (https://github.com/envoyproxy/envoy/blob/06505f2424ad3f894af404a07bcfd7582aa8f3b9/test/integration/tcp_tunneling_integration_test.cc#L64).  This is set to "host:80" by default, which is not resolvable by DFP DNS, will cause connection creation fail, hence test fail.  The removeHost()/setHost() here is to change the :authority filed into a locally resolvable domain name, i.e, 'localhost' to make it work.

There are existing integration test code which using 'localhost' for DFP filter to perform DNS resolution: https://github.com/envoyproxy/envoy/blob/06505f2424ad3f894af404a07bcfd7582aa8f3b9/test/extensions/filters/http/dynamic_forward_proxy/proxy_filter_integration_test.cc#L225.